File: 1884-minimum-changes-to-make-alternating-binary-string/minimum-changes-to-make-alternating-binary-string.java
class Solution {
    public int minOperations(String s) {
        return Math.min(startsWithZero(s), startsWithOne(s));
    }

    private int startsWithZero(String s) {
        int cur = '0';
        int diffCount = 0;
        for (char c : s.toCharArray()) {
            if (c != cur) {
                diffCount++;
            }
            cur = cur == '0' ? '1' : '0';
        }
        return diffCount;
    }
    
    private int startsWithOne(String s) {
        int cur = '1';
        int diffCount = 0;
        for (char c : s.toCharArray()) {
            if (c != cur) {
                diffCount++;
            }
            cur = cur == '0' ? '1' : '0';
        }
        return diffCount;
    }
}

File: 1884-minimum-changes-to-make-alternating-binary-string/minimum-changes-to-make-alternating-binary-string.py
class Solution:
    def minOperations(self, s: str) -> int:
        keep_first_digit_ops = 0
        cur = s[0]
        for d in s[1:]:
            cur = '0' if cur == '1' else '1'
            # print(f'd={d} cur={cur}')
            if d != cur:
                keep_first_digit_ops += 1
        # print(f'keep_first_digit_ops={keep_first_digit_ops}')
        change_first_digit_ops = 1
        cur = '0' if s[0] == '1' else '1'
        for d in s[1:]:
            cur = '0' if cur == '1' else '1'
            print(f'd={d} cur={cur}')
            if d != cur:
                change_first_digit_ops += 1
        # print(f'change_first_digit_ops={change_first_digit_ops}')
        return min(keep_first_digit_ops, change_first_digit_ops)